src/runtime/memory/mod.rs

```diff
@@ -27,111 +28,128 @@ pub use self::{buffer_pool::*, demibuffer::*};
 //======================================================================================================================
 
 pub trait DemiMemoryAllocator {
+    fn get_max_buffer_size_bytes(&self) -> usize {
+        u16::MAX as usize
+    }
+
     fn allocate_demi_buffer(&self, size: usize) -> Result<DemiBuffer, Fail> {
         Ok(DemiBuffer::new(size as u16))
     }
 }
 
-/// Converts a buffer into a scatter-gather array.
-pub fn into_sgarray(buf: DemiBuffer) -> Result<demi_sgarray_t, Fail> {
-    // Create a scatter-gather segment to expose the DemiBuffer to the user.
-    let data: *const u8 = buf.as_ptr();
-    let sga_seg: demi_sgaseg_t = demi_sgaseg_t {
-        sgaseg_buf: data as *mut c_void,
-        sgaseg_len: buf.len() as u32,
-    };
+/// Converts a list of DemiBuffers into a scatter-gather array.
+pub fn into_sgarray(bufs: ArrayVec<DemiBuffer, DEMI_SGARRAY_MAXLEN>) -> Result<demi_sgarray_t, Fail> {
+    // Check the sizes before allocating anything.
+    if bufs.is_empty() {
+        let cause: String = format!("cannot allocate a zero element scatter-gather array");
```

Comment:
A string literal can be used here instead of `format!` to make it more efficient and readable, also there will be no need to take the reference to `cause` with that change. The block can be written like this:

```suggestion
        let cause: &'static str = "cannot allocate a zero element scatter-gather array";
        error!("into_sgarray(): {}", cause);
        return Err(Fail::new(libc::EINVAL, cause));
```